README.md
**Project name:**
PV simulator challenge

**Description:**

This project is a challenge to simulate simple PV systems. It has 3 main parts:
- consumer's meter
- broker
- pv simulator

1. Meter.

Its simple application to simulate power meter of consumer. 
Meter application makes simulation of power consumption of consumer and send it to broker.

2. Broker.

I'm  using RabbitMQ to communicate with other applications. It received messages from meter(s) and send them to pv simulator(s).

3. PV simulator.

There is simple simulator of PV system. It receives meter's value from broker and simulate PV system. All information from PV system stored to output file.

**How to run:**

You should use Docker to run this project. 
1. Install and run Docker: https://docs.docker.com/engine/install/
2. Build and run project: docker-compose -f docker-compose.yml up --build
3. You can see output of "meter" and "pv simulator" in ./data/logs/ directory.
4. Result output is in ./data/log/ directory.

**Configurations**:

All required parameters set in docker-compose.yml file. For "DEV" mode you can use docker-compose-dev.yml file.
- BROKER_HOST  - hostname of broker (RabbitMQ)
- BROKER_PORT  - port of broker (RabbitMQ)
- BROKER_QUEUE - queue name of broker (RabbitMQ)
- BROKER_USERNAME - username of broker (RabbitMQ)
- BROKER_PASSWORD - password of broker (RabbitMQ)
- PV_MIN - minimal power of meter (W)
- PV_MAX - maximal power of meter (W)
- ITERATION - interval where sending values from meter (s)
- LOGFILE - name of log file
- DELIMITER - delimiter of log file
- OUTPUT_FILE - name of output file
- TIME_ITER - value is spacing between two data simulations (s)
- MAX_EXECUTE_TIME - if simulation time is more than this value, it will be noted in log file (s)
- EXECUTE_TIME_LOG - log file name for execute time of each daily simulation 
- PV_SUNRISE_START - sunrise start time (h) (got value from chart from main "homework" file)
- PV_SUNRISE_END - sunrise end time (h) (got value from chart from main "homework" file)
- PV_ZENITH - zenith time (h) (got value from chart from main "homework" file)
- PV_SUNDOWN_START - sundown start time (h) (got value from chart from main "homework" file)
- PV_SUNDOWN_END - sundown end time (h) (got value from chart from main "homework" file)
- PV_LIGHT_EFF_LW - light efficiency of light in sunrise and sunset time
- PV_LIGHT_EFF_STD - light efficiency of standard time
- PV_MAX_POWER - maximal power of PV system (W)

**Testing:**

Start project with "docker-compose -f docker-compose-dev.yml up --build" and run unit tests of each application.
Meter - "docker exec -ti meter python -m pytest"
PV Simulator - "docker exec -ti pv python -m pytest"




CHANGELOG
# Version 1.1 Changes
1. Changed "Broker startup delay" procedure
Now app didn't waiting to 20 seconds when the broker starts. App checking broker's open port and when port starts to listening, app starts simulation.
Checking thru netcat utility from entrypoint bash script.

2. Changed from "realtime simulation" to once per minute simulation.
Now app didn't waiting (sleeping) for next simulation iteration and generating new data without delay.
TIME_ITER (sec.) value is spacing between two data simulations.
QTY of simulations per day is 24*60*60/TIME_ITER.

3. Changed output csv procedure.
Now to output file saves data who contains daily values,
so app generating several output files in format filename_dayXXX.csv.

4. Added execution time checker of each simulating day.
Meashures execution time from starting meter simulator up to saving final value to output file.
Execution time is saved to file named "execution_time.log".

5. Changed algorithm of generating meter's value and pv simulator's value.

6. Added ploting of data from output file.

7. For testing I used pytest framework.
Added unit tests for new functionality.

8. Added "DEV" mode, so for testing should start "docker-compose -f docker-compose-dev.yml up --build"
I've added to DEV mode pytest dependencies.






services/pv/tests/test_PV.py
from datetime import datetime

import pika
from pv.PV import Pv
import unittest
import logging


class testPv(unittest.TestCase, Pv):
    """
    Class for testing PV Simulator
    """
    _environment_pv = "TEST"
    _broker_host = "rabbitmq"
    _broker_port = "5672"
    _broker_queue = "meter_simulator_test"
    _broker_username = "guest"
    _broker_password = "guest"
    _output_file = "./log/output_test.csv"
    _delimiter = ";"
    _pv_min = 0
    _pv_max = 9000
    _logfile = "./log/pv_log.log"
    _time_iter = 60
    _max_execute_time = 60
    _execute_time_log = "./log/execute_time_test.log"
    _pv_sunrise_start = 6
    _pv_sunrise_end = 8
    _pv_zenith = 14
    _pv_sundown_start = 20
    _pv_sundown_end = 21
    _pv_light_eff_lw = 0.1
    _pv_light_eff_std = 0.8125
    _pv_max_power = 9000

    _credentials = pika.PlainCredentials(_broker_username, _broker_password)

    def test_connection(self):
        """
        Test connection to broker.
        :return:
        """
        try:
            self._connect_broker()
            self.assertTrue(True)
        except Exception:
            self.assertTrue(False)

    def test_generating(self):
        """
        Test generating pv simulator's value.
        :return:
        """
        try:
            self._generate_pv_value(1, 1)
            self.assertTrue(True)
        except Exception:
            self.assertTrue(False)

    def test_writing(self):
        """
        Test writing pv simulator's value to output file.
        :return:
        """
        try:
            data_record = {
                "timestamp": 1,
                "meter_value": "-1234",
                "pv_value": "1234",
                "sum": "0"
            }
            self._write_to_output(1, data_record, "DATA")
            self.assertTrue(True)
        except Exception:
            self.assertTrue(False)

    def test_get_value_from_broker(self):
        try:
            channel = self._connect_broker()
            self._get_value_from_broker(channel)
            self.assertTrue(True)
        except Exception:
            self.assertTrue(False)

    def test_parse_data_string(self):
        assert self._parse_data_string("1;1;1234") == [1, 1, 1234]

    def test_make_filename(self):
        assert self._make_filename(1) == "./log/output_test_day1.csv"

    def test_make_plot_filename(self):
        assert self._make_plot_filename(1) == "./log/output_test_day1.png"

    def test_get_fraction_time(self):
        assert self._get_fraction_time(0) == 0
        assert self._get_fraction_time(24) == 1

    def test_morning_strgt(self):
        assert self._morning_strgt(0) == -2700.0

    def test_evening_strgt(self):
        assert self._evening_strgt(1) == 144000.0

    def test_parabola(self):
        assert self._parabola(0) == -17800.0


if __name__ == "__main__":
    unittest.main()


services/meter/tests/test_Meter.py
import pika
from meter.Meter import Meter
import unittest
import logging


class testMeter(unittest.TestCase, Meter):
    """
    Class for testing Meter.
    """
    _environment_pv = "TEST"
    _broker_host = "rabbitmq"
    _broker_port = '5672'
    _broker_queue = "meter_simulator_test"
    _broker_username = "guest"
    _broker_password = "guest"
    _pv_min = 0
    _pv_max = 9000
    _time_iter = 60
    _logfile = "./log/tests.log"
    _max_consume = 14

    _credentials = pika.PlainCredentials(_broker_username, _broker_password)

    def test_connection(self):
        """
        Test connection to broker.
        :return:
        """
        try:
            self._connect_broker()
            self.assertTrue(True)
        except Exception:
            self.assertTrue(False)

    def test_publishing(self):
        """
        Test publishing meter's value to broker.
        :return:
        """
        try:
            channel = self._connect_broker()
            self._publish_meter_to_broker(channel, "1234")
            self.assertTrue(True)
        except Exception:
            self.assertTrue(False)

    def test_generating(self):
        """
        Test generating meter's value.
        :return:
        """
        try:
            self._generate_meter(2)
            self.assertTrue(True)
        except Exception:
            self.assertTrue(False)

    def test_morning_strgt_meter(self):
        assert self._morning_strgt_meter(0) == 0

    def test_evening_strgt_meter(self):
        assert self._evening_strgt_meter(1) == 0

    def test_get_fraction_time(self):
        assert self._get_fraction_time(0) == 0
        assert self._get_fraction_time(24) == 1


if __name__ == "__main__":
    unittest.main()


services/pv/PV.py
import csv
import logging
import os
from datetime import datetime
from random import randint
from typing import List

import matplotlib.pyplot as plt
import pandas as pd
import pika


class Pv:
    """
    Class receiving consumer's meter value from a broker, generating PV value,
    writing to a file and makes plot of output data.
    Should provide broker's host, port, queue, username, password.
    """

    def __init__(self, broker_host: str, broker_port: int, broker_queue: str,
                 broker_username: str,
                 broker_password: str,
                 pv_min: int, pv_max: int,
                 output_file: str, delimiter: str, logfile: str,
                 environment_pv: str, max_execute_time: int, time_iter: int,
                 execute_time_log: str, pv_sunrise_start: int,
                 pv_sunrise_end: int, pv_zenith: int, pv_sundown_start: int,
                 pv_sundown_end: int, pv_light_eff_lw: int,
                 pv_light_eff_std: int, pv_max_power: int):
        self._logfile = logfile
        self._broker_host = broker_host
        self._broker_port = broker_port
        self._broker_queue = broker_queue
        self._broker_username = broker_username
        self._broker_password = broker_password
        self._pv_min = pv_min
        self._pv_max = pv_max
        self._output_file = output_file
        self._delimiter = delimiter
        self._max_execute_time = max_execute_time
        self._time_iter = time_iter
        self._execute_time_log = execute_time_log
        self._start_time = 0
        self._credentials = pika.PlainCredentials(self._broker_username,
                                                  self._broker_password)
        self._environment_pv = environment_pv
        self._pv_sunrise_start = self._get_fraction_time(pv_sunrise_start)
        self._pv_sunrise_end = self._get_fraction_time(pv_sunrise_end)
        self._pv_zenith = self._get_fraction_time(pv_zenith)
        self._pv_sundown_start = self._get_fraction_time(pv_sundown_start)
        self._pv_sundown_end = self._get_fraction_time(pv_sundown_end)
        self._pv_light_eff_lw = pv_light_eff_lw
        self._pv_light_eff_std = pv_light_eff_std
        self._pv_max_power = pv_max_power
        logging.basicConfig(filename=logfile, filemode="a", level=logging.INFO,
                            format='%(asctime)s %(message)s',
                            datefmt='%Y-%m-%d %H:%M:%S %p')

    def _connect_broker(self):
        """
        Connects to broker. Returns channel.
        :return:
        """
        logging.info("PV: Connecting to broker")
        try:
            self._connection = pika.BlockingConnection(
                pika.ConnectionParameters(self._broker_host,
                                          port=self._broker_port,
                                          credentials=self._credentials))
            channel = self._connection.channel()
            channel.basic_qos(prefetch_count=1)
            channel.queue_declare(queue=self._broker_queue, durable=True)
            return channel
        except pika.exceptions.ConnectionClosedByBroker as e:
            logging.error("PV: Connection to broker closed by broker: %s", e)
        except pika.exceptions.AMQPConnectionError as e:
            logging.error("PV: Connection to broker failed: %s", e)
        except Exception as e:
            logging.error("PV: Error: %s", e)
        else:
            logging.info("PV: Connected to broker")
        return None

    def start(self):
        """
        Connects to broker and receives meter's value.
        :return:
        """
        logging.info("PV: Starting PV")
        channel = self._connect_broker()
        if channel is not None:
            self._get_value_from_broker(channel)

    def _get_value_from_broker(self, channel: pika.channel.Channel):
        try:
            channel.basic_consume(queue=self._broker_queue,
                                  on_message_callback=self._callback)
            channel.start_consuming()
        except pika.exceptions.ConnectionClosedByBroker as e:
            logging.error("PV: Connection to broker closed by broker: %s", e)
        except pika.exceptions.AMQPConnectionError as e:
            logging.error("PV: Connection to broker failed: %s", e)
        except KeyboardInterrupt as e:
            channel.stop_consuming()
            self._connection.close()
            logging.info("PV: Stopped PV")
        except Exception as e:
            logging.error("PV: Error: %s", e)

    def _callback(self, ch, method, properties, body):
        """
        Callback function for receiving meter's value.
        """
        logging.info(f"PV: Received message {body}")

        data_from_broker = body.decode("utf-8")
        command_meter, value = data_from_broker.split("::")
        if command_meter == "START":
            self._start_time = int(value)
        elif command_meter == "DATA":
            meter_day, meter_iteration, meter_reading = self._parse_data_string(
                value)
            if meter_reading < self._pv_min or meter_reading > self._pv_max:
                logging.info(
                    f"PV: Meter reading {meter_reading} is out of range")
            else:
                pv_value = self._generate_pv_value(meter_day, meter_iteration)
                data_record = {
                    "timestamp": meter_iteration,
                    "meter_value": meter_reading * -1,
                    "pv_value": pv_value,
                    "sum": pv_value - meter_reading,
                }
                logging.info(f"PV: Writing data record {data_record} to file")
                self._write_to_output(meter_day, data_record, command_meter)
        elif command_meter == "END":
            meter_day, meter_iteration, meter_reading = self._parse_data_string(
                value)
            filename = self._make_filename(meter_day)
            logging.info(f"PV: Renaming temporary file to {filename}")
            os.rename(filename + ".tmp", filename)
            self._make_plot(meter_day)
            data_file = open(self._execute_time_log, 'a')

            execution_time = int(datetime.now().timestamp()) - int(
                self._start_time)
            data_file.write(
                f"Execution time of {filename}: {execution_time} seconds\n")
            if execution_time > self._max_execute_time:
                logging.info(
                    f"PV: Execution time of {filename} is too long:"\
                        f"{execution_time} seconds")
            data_file.close()
        ch.basic_ack(delivery_tag=method.delivery_tag)

    def _make_plot(self, meter_day: int):
        """
        Drawing a plot of daily data.
        :param meter_day:
        :return:
        """
        filename = self._make_filename(meter_day)
        plot_filename = self._make_plot_filename(meter_day)
        logging.info(f"PV: Making plot to {plot_filename}")
        try:
            data_csv = pd.read_csv(filename, delimiter=";",
                                   index_col='timestamp')
            data_csv.plot()
            plt.savefig(plot_filename)
        except Exception as e:
            logging.error(f"PV: Can't plot data to {plot_filename}! Error: {e}")

    def _parse_data_string(self, data_string: str) -> List:
        """
        Parses data from broker and returns day, timestamp,
        generated value from meter.
        """
        meter_day, meter_iteration, meter_reading = data_string.split(
            self._delimiter)
        return [int(meter_day), int(meter_iteration), int(meter_reading)]

    def _make_filename(self, meter_day: int) -> str:
        """
        Makes filename for data file.
        :return:
        """
        if ".csv" in self._output_file:
            return f"{self._output_file.split('.csv')[0]}_day{meter_day}.csv"
        return f"day{meter_day}_{self._output_file}"

    def _make_plot_filename(self, meter_day: int) -> str:
        """
        Makes filename for plot file.
        :return:
        """
        if ".csv" in self._output_file:
            return f"{self._output_file.split('.csv')[0]}_day{meter_day}.png"
        return f"day{meter_day}_{self._output_file}.png"

    def _write_to_output(self, meter_day: int, data_record: dict,
                         command_meter: str):
        """
        Writes data record to output file.
        :return:
        """
        filename = self._make_filename(meter_day) + ".tmp"
        new_day = False
        if not os.path.isfile(filename):
            new_day = True
        data_file = open(filename, 'a')
        if command_meter == "DATA":
            csv_writer = csv.writer(data_file, delimiter=self._delimiter)
            if new_day:
                csv_writer.writerow(["timestamp", "meter", "pv", "sum"])
            csv_writer.writerow(data_record.values())
        data_file.close()

    def _get_fraction_time(self, hour: int, minute: int = 0,
                           second: int = 0) -> float:
        """Convert time to fraction of the day"""
        return ((hour * 3600 + minute * 60) + second) / (24 * 3600)

    def _get_parabola_vertex(self, x1: float, y1: float, x2: float, y2: float,
                             x3: float,
                             y3: float) -> (float, float, float):
        """Calculate the vertex of a parabola given three points"""

        denom = (x1 - x2) * (x1 - x3) * (x2 - x3)
        A = (x3 * (y2 - y1) + x2 * (y1 - y3) + x1 * (y3 - y2)) / denom
        B = (x3 * x3 * (y1 - y2) +
             x2 * x2 * (y3 - y1) +
             x1 * x1 * (y2 - y3)) / denom
        C = (x2 * x3 * (x2 - x3) * y1 +
             x3 * x1 * (x3 - x1) * y2 +
             x1 * x2 * (x1 - x2) * y3) / denom

        return A, B, C

    def _morning_strgt(self, x: float) -> float:
        A = (self._pv_light_eff_lw * self._pv_max_power) / (
                self._pv_sunrise_end - self._pv_sunrise_start)
        B = A * self._pv_sunrise_start * -1
        return A * x + B

    def _evening_strgt(self, x: float) -> float:
        A = ((-self._pv_sunrise_end) * self._pv_max_power *
             self._pv_light_eff_lw) / (self._pv_sundown_end -
                                       self._pv_sundown_start)
        B = A * self._pv_sundown_end * -1
        return A * x + B

    def _parabola(self, x: float) -> float:
        """Calculate the parabola"""
        (A, B, C) = self._get_parabola_vertex(
            self._pv_sunrise_end, self._morning_strgt(self._pv_sunrise_end),
            self._pv_zenith, (self._pv_light_eff_std * self._pv_max_power),
            self._pv_sundown_start, self._evening_strgt(self._pv_sundown_start))
        return A * (x ** 2) + (B * x) + C

    def _generate_pv_value(self, meter_day: int, meter_iteration: int) -> int:
        """Generate PV curve using a straight for SUNRISE
        a parabola for ZENITH and another straight for SUNDOWN"""

        time = self._get_fraction_time(0, 0, meter_iteration * self._time_iter)

        if time >= self._pv_sunrise_start and time <= self._pv_sunrise_end:
            value = self._morning_strgt(time)
        elif time > self._pv_sunrise_end and time < self._pv_sundown_start:
            value = self._parabola(time)
        elif time >= self._pv_sundown_start and time <= self._pv_sundown_end:
            value = self._evening_strgt(time)
        else:
            value = 0.0

        return int(value)


services/meter/Meter.py
import logging
import time
from datetime import datetime
from random import randint

import pika


class Meter:
    """
    Meter class simulates consumer meter. Class generates value between
    pv_min and pv_max (Watt) and publishes value to broker.
    Should provide broker's host, port, queue, username, password.
    Class publishes meter's value every time_iter seconds.
    """

    def __init__(self, broker_host: str, broker_port: int, broker_queue: str,
                 broker_username: str,
                 broker_password: str, pv_min: int, pv_max: int,
                 time_iter: int, logfile: str, environment_pv: str,
                 delimiter: str, max_consume: int):

        self._logfile = logfile
        self._broker_host = broker_host
        self._broker_port = broker_port
        self._broker_queue = broker_queue
        self._broker_username = broker_username
        self._broker_password = broker_password
        self._pv_min = pv_min
        self._pv_max = pv_max
        self._time_iter = time_iter
        self._total_iterations = 24 * 60 * 60 / self._time_iter
        self._credentials = pika.PlainCredentials(self._broker_username,
                                                  self._broker_password)
        self._environment_pv = environment_pv
        self._delimiter = delimiter
        self._max_consume = max_consume
        logging.basicConfig(filename=logfile, filemode="a", level=logging.INFO,
                            format='%(asctime)s %(message)s',
                            datefmt='%Y-%m-%d %H:%M:%S %p')

    def _connect_broker(self):
        """
        Connects to broker and returns channel.
        """
        logging.info("Meter: Connecting to broker: %s", self._broker_host)
        try:
            self._connection = pika.BlockingConnection(
                pika.ConnectionParameters(host=self._broker_host,
                                          port=self._broker_port,
                                          credentials=self._credentials))
            channel = self._connection.channel()
            channel.queue_declare(queue=self._broker_queue, durable=True)
            channel.confirm_delivery()
            return channel
        except pika.exceptions.ConnectionClosedByBroker as e:
            logging.error("Meter: Connection to broker closed by broker: %s", e)
        except pika.exceptions.AMQPConnectionError as e:
            logging.error("Meter: Connection to broker failed: %s", e)
        except Exception as e:
            logging.error("Meter: Error: %s", e)
        return None

    def start(self):
        """
        Coneects to broker and starts meter simulator.
        """
        logging.info("Meter: Connecting to broker: %s", self._broker_host)
        channel = self._connect_broker()
        if channel is not None:
            self._current_day = 0

            while True:
                data_to_send = f"START::{int(datetime.now().timestamp())}"
                try:
                    self._publish_meter_to_broker(channel, data_to_send)
                except KeyboardInterrupt:
                    logging.info("Meter: Exiting meter simulator")
                    break
                self._current_iteration = 1
                while self._current_iteration <= self._total_iterations:
                    try:
                        meter = self._generate_meter(self._current_iteration)
                        data_to_send = f"DATA::"\
                                f"{self._make_string_to_broker(meter)}"
                        self._publish_meter_to_broker(channel, data_to_send)
                        self._current_iteration += 1
                    except KeyboardInterrupt:
                        logging.info("Meter: Exiting meter simulator")
                        break
                data_to_send = f"END::{self._make_string_to_broker(0)}"
                try:
                    self._publish_meter_to_broker(channel, data_to_send)
                except KeyboardInterrupt:
                    logging.info("Meter: Exiting meter simulator")
                    break
                self._current_day += 1
                time.sleep(3)
        self._connection.close()

    def _generate_meter(self, current_iteration: int) -> int:
        """
        Generates value between pv_min and pv_max (Watt)
        """
        time = self._get_fraction_time(0, 0,
                                       current_iteration * self._time_iter)
        if current_iteration <= self._max_consume * 60:
            meter_value = int(self._morning_strgt_meter(time))
        else:
            meter_value = int(self._evening_strgt_meter(time))
        if meter_value < self._pv_min:
            meter_value = self._pv_min
        elif meter_value > self._pv_max:
            meter_value = self._pv_max
        return meter_value

    def _morning_strgt_meter(self, x: float) -> float:
        A = self._pv_max * 0.75 / (self._get_fraction_time(
            randint(self._max_consume - 2,
                    self._max_consume + 2)))
        return A * x

    def _evening_strgt_meter(self, x: float) -> float:
        A = (-self._pv_max * 0.75) / (
                1 - self._get_fraction_time(
            randint(self._max_consume - 2, self._max_consume + 2)))
        B = A * -1
        return A * x + B

    def _get_fraction_time(self, hour: int, minute: int = 0,
                           second: int = 0) -> float:
        """Convert time to fraction of the day"""
        return ((hour * 3600 + minute * 60) + second) / (24 * 3600)

    def _make_string_to_broker(self, meter: int) -> str:

        return f"{self._current_day}{self._delimiter}" \
               f"{self._current_iteration}{self._delimiter}" \
               f"{meter}"

    def _publish_meter_to_broker(self, channel, data_to_send: str):
        """
        Publishes simulated value to broker.
        """
        logging.info("Meter: Publishing meter to broker: %s", data_to_send)
        try:
            channel.basic_publish(exchange='', routing_key=self._broker_queue,
                                  body=data_to_send,
                                  properties=pika.BasicProperties(
                                      delivery_mode=2))
        except pika.exceptions.ConnectionClosedByBroker as e:
            logging.error("Meter: Connection to broker closed by broker: %s", e)
        except pika.exceptions.AMQPConnectionError as e:
            logging.error("Meter: Connection to broker failed: %s", e)
        except Exception as e:
            logging.error("Meter: Error while publishing PV: %s", e)


services/pv/__main__.py
import logging
import os
import time
from datetime import datetime
from functools import wraps

from pv.PV import Pv


def main():
    """
    Initializes PV and starts PV simulator.
    :return:
    """
    BROKER_HOST = os.environ.get('BROKER_HOST', 'rabbitmq')
    BROKER_PORT = os.environ.get('BROKER_PORT', '5672')
    BROKER_QUEUE = os.environ.get('BROKER_QUEUE', 'meter_simulator')
    BROKER_USERNAME = os.environ.get('BROKER_USERNAME', 'guest')
    BROKER_PASSWORD = os.environ.get('BROKER_PASSWORD', 'guest')
    PV_MIN = int(os.environ.get('PV_MIN', 0))
    PV_MAX = int(os.environ.get('PV_MAX', 9000))
    OUTPUT_FILE = os.environ.get('OUTPUT_FILE', 'output.csv')
    DELIMITER = os.environ.get('DELIMITER', ';')
    LOGFILE = os.environ.get('LOGFILE', './log/pv.log')
    ENVIRONMENT_PV = os.environ.get('ENVIRONMENT_PV', 'DEV')
    MAX_EXECUTE_TIME = int(os.environ.get('MAX_EXECUTE_TIME', 60))
    TIME_ITER = int(os.environ.get('TIME_ITER', 60))
    EXECUTE_TIME_LOG = os.environ.get('EXECUTE_TIME_LOG',
                                      './log/execute_time.log')
    PV_SUNRISE_START = int(os.environ.get('PV_SUNRISE_START', 6))
    PV_SUNRISE_END = int(os.environ.get('PV_SUNRISE_END', 8))
    PV_ZENITH = int(os.environ.get('PV_ZENITH', 14))
    PV_SUNDOWN_START = int(os.environ.get('PV_SUNDOWN_START', 20))
    PV_SUNDOWN_END = int(os.environ.get('PV_SUNDOWN_END', 21))
    PV_LIGHT_EFF_LW = float(os.environ.get('PV_LIGHT_EFF_LW', 0.1))
    PV_LIGHT_EFF_STD = float(os.environ.get('PV_LIGHT_EFF_STD', 0.8125))
    PV_MAX_POWER = int(os.environ.get('PV_MAX_POWER', 4000))

    pv = Pv(BROKER_HOST, BROKER_PORT, BROKER_QUEUE, BROKER_USERNAME,
            BROKER_PASSWORD, PV_MIN, PV_MAX, OUTPUT_FILE,
            DELIMITER, LOGFILE, ENVIRONMENT_PV, MAX_EXECUTE_TIME, TIME_ITER,
            EXECUTE_TIME_LOG, PV_SUNRISE_START, PV_SUNRISE_END, PV_ZENITH,
            PV_SUNDOWN_START, PV_SUNDOWN_END, PV_LIGHT_EFF_LW, PV_LIGHT_EFF_STD,
            PV_MAX_POWER)
    pv.start()


if __name__ == '__main__':
    main()
